lib/testutils/src/lib.rs

```diff
@@ -383,61 +383,142 @@ pub fn write_file(store: &Store, path: &RepoPath, contents: &str) -> FileId {
         .unwrap()
 }
 
-pub fn write_normal_file(
-    tree_builder: &mut TreeBuilder,
-    path: &RepoPath,
-    contents: &str,
-) -> FileId {
-    let id = write_file(tree_builder.store(), path, contents);
-    tree_builder.set(
-        path.to_owned(),
-        TreeValue::File {
-            id: id.clone(),
+pub struct TestTreeBuilder {
+    store: Arc<Store>,
+    tree_builder: TreeBuilder,
+}
+
+impl TestTreeBuilder {
+    pub fn new(store: Arc<Store>) -> Self {
+        let tree_builder = store.tree_builder(store.empty_tree_id().clone());
+        Self {
+            store,
+            tree_builder,
+        }
+    }
+
+    pub fn binary_file(
+        &mut self,
+        path: &RepoPath,
+        contents: Vec<u8>,
+    ) -> TestTreeFileEntryBuilder<'_> {
+        TestTreeFileEntryBuilder {
+            tree_builder: &mut self.tree_builder,
+            path: path.to_owned(),
+            contents,
             executable: false,
-        },
-    );
-    id
+        }
+    }
+
+    pub fn text_file(&mut self, path: &RepoPath, contents: &str) -> TestTreeFileEntryBuilder<'_> {
```

Comment:
nit: `contents` can be `AsRef<[u8]>` or `Into<Vec<u8>>`. Then, we can merge `binary`/`text_file()` functions, which should be identical. `empty_file()` can also be removed, but I don't have strong feeling about that.